go/capship/commands/main.go

```diff
@@ -0,0 +1,117 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"context"
+	"os"
+	"os/signal"
+
+	"github.com/IBM/cap/go/log"
+	"github.com/IBM/cap/go/version"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+const usage = `
+                                .__     .__
+  ____  _____   ______    ______|  |__  |__|______
+_/ ___\ \__  \  \____ \  /  ___/|  |  \ |  |\____ \
+\  \___  / __ \_|  |_> > \___ \ |   Y  \|  ||  |_> >
+ \___  >(____  /|   __/ /____  >|___|  /|__||   __/
+     \/      \/ |__|         \/      \/     |__|
+
+a server for CAP Alerts and Atomfeeds with CAP Alert Summaries
+`
+
+// App returns a *cli.App instance.
+func App() *cli.App {
+	app := cli.NewApp()
+	app.Name = "capship"
+	app.Version = version.Version
+	app.Usage = usage
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "config,c",
+			Usage: "path to the configuration file",
+			Value: DefaultConfigPath,
+		},
+		cli.StringFlag{
+			Name:  "root",
+			Usage: "capship root directory",
+		},
+		cli.Int64Flag{
+			Name:  "max-upload-size,m",
+			Usage: "int64",
+			Value: DefaultMaxUploadSize,
+		},
+		cli.StringFlag{
+			Name:  "log-level,l",
+			Usage: "logging level [trace, debug, info, warn, error, fatal, panic]",
+		},
+	}
+	app.Commands = []cli.Command{
+		configCommand,
+	}
+	app.Action = func(c *cli.Context) error {
+		var (
+			signalsCH = make(chan os.Signal, 2048)
+			serverCH  = make(chan *Server, 1)
+			bctx      = context.Background()
+			config    = defaultConfig()
+		)
+
+		done := handleSignals(bctx, signalsCH, serverCH)
+		signal.Notify(signalsCH, handledSignals...)
+
+		if err := LoadConfig(c.GlobalString("config"), config); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		// apply flags to the config
+		if err := applyFlags(c, config); err != nil {
+			return err
+		}
+
+		log.G(bctx).WithFields(logrus.Fields{
+			"version":  version.Version,
+			"revision": version.Revision,
+		}).Info("starting capship")
+
+		server, err := New(bctx, config)
+		if err != nil {
+			return err
+		}
+		log.G(bctx).WithFields(logrus.Fields{
+			"log-level": log.PrintLevel(log.G(bctx).Logger),
+		}).Info("logger")
+
+		serverCH <- server
+		go server.serve()
+
+		log.G(bctx).Infof("capship ready:")
+		log.G(bctx).Infof("   use '/cap/' to pull the cap alert feed")
+		log.G(bctx).Infof("   use '/cap/{reference}' to pull a cap alert file")
+		log.G(bctx).Infof("   use '/upload' to upload unique alert files using curl etc. Example:")
+		log.G(bctx).Infof("      $ curl -F 'uploadFile=@KAR0-0306112239-SW.xml' http://localhost:8080/upload")
+		log.G(bctx).Infof("   use '/feeds/{fileName}' to download feed files")
+		log.G(bctx).Infof("   use '/alerts/{fileName}' to download alert files")
```

Comment:
already not a fan of this logging system, but okay for now.

Reply:
the output is nice... and in color
```
INFO[0000] starting capship                              revision=fa6074923dec08fdbe204d7558d1374dca75d8fe version=fa60749
INFO[0000] logger                                        log-level=trace
INFO[0000] capship ready serving on: http://localhost:8080/ 
INFO[0000]    use '/cap/' to pull the cap alert feed    
INFO[0000]    use '/cap/{reference}' to pull a cap alert file 
INFO[0000]    use '/upload' to upload unique alert files using curl etc. Example: 
INFO[0000]       $ curl -F 'uploadFile=@KAR0-0306112239-SW.xml' http://localhost:8080/upload 
INFO[0000]    use '/feeds/{fileName}' to download feed files 
INFO[0000]    use '/alerts/{fileName}' to download alert files 
DEBU[0733] received signal                               signal="user defined signal 1"
INFO[0733] === BEGIN stack dump of all goroutines ===
goroutine 20 [running]:
github.com/IBM/cap/go/capship/commands.dumpStacks()
	/home/mike/go/src/github.com/IBM/cap/go/capship/commands/signals.go:69 +0xa7
github.com/IBM/cap/go/capship/commands.handleSignals.func1(0xc00008a540, 0xc00008a4e0, 0x85c900, 0xc000084038, 0xc000068240)
	/home/mike/go/src/github.com/IBM/cap/go/capship/commands/signals.go:48 +0x24f
created by github.com/IBM/cap/go/capship/commands.handleSignals
	/home/mike/go/src/github.com/IBM/cap/go/capship/commands/signals.go:38 +0x89

goroutine 1 [chan receive]:
github.com/IBM/cap/go/capship/commands.App.func1(0xc0000a6580, 0xc0000a6580, 0xc0000efd17)
	/home/mike/go/src/github.com/IBM/cap/go/capship/commands/main.go:129 +0x9d3
github.com/IBM/cap/vendor/github.com/urfave/cli.HandleAction(0x776560, 0x80b898, 0xc0000a6580, 0xc00008a480, 0x0)
	/home/mike/go/src/github.com/IBM/cap/vendor/github.com/urfave/cli/app.go:501 +0xc8
github.com/IBM/cap/vendor/github.com/urfave/cli.(*App).Run(0xc00011c000, 0xc000078190, 0x1, 0x1, 0x0, 0x0)
	/home/mike/go/src/github.com/IBM/cap/vendor/github.com/urfave/cli/app.go:268 +0x5b7
main.main()
	/home/mike/go/src/github.com/IBM/cap/go/capship/main.go:28 +0x4f

goroutine 19 [syscall]:
os/signal.signal_recv(0x85bda0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

goroutine 4 [IO wait]:
internal/poll.runtime_pollWait(0x7f9659c87f00, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc00013c218, 0x72, 0xc00008c000, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc00013c218, 0xffffffffffffff00, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc00013c200, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:384 +0x1a0
net.(*netFD).accept(0xc00013c200, 0x50, 0x7b7700, 0xc000049c01)
	/usr/local/go/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc00000e018, 0xc000049c40, 0xc0000a02d0, 0x2c2661af38c030a3)
	/usr/local/go/src/net/tcpsock_posix.go:139 +0x2e
net.(*TCPListener).AcceptTCP(0xc00000e018, 0xc000049c68, 0x485096, 0x5cfeaef9)
	/usr/local/go/src/net/tcpsock.go:247 +0x47
net/http.tcpKeepAliveListener.Accept(0xc00000e018, 0xc000049cb8, 0x18, 0xc000001500, 0x6f9125)
	/usr/local/go/src/net/http/server.go:3232 +0x2f
net/http.(*Server).Serve(0xc000132270, 0x85c800, 0xc00000e018, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:2826 +0x22f
net/http.(*Server).ListenAndServe(0xc000132270, 0xc000132270, 0xc000136270)
	/usr/local/go/src/net/http/server.go:2764 +0xb6
net/http.ListenAndServe(0x7f2383, 0x5, 0x0, 0x0, 0x8599e0, 0xc000136270)
	/usr/local/go/src/net/http/server.go:3004 +0x74
github.com/IBM/cap/go/capship/commands.(*Server).serve(0xc00000c180)
	/home/mike/go/src/github.com/IBM/cap/go/capship/commands/server.go:81 +0x31f
created by github.com/IBM/cap/go/capship/commands.App.func1
	/home/mike/go/src/github.com/IBM/cap/go/capship/commands/main.go:119 +0x69f

=== END stack dump === 
```